### PR TITLE
export record locks

### PR DESCRIPTION
**Pull request by teallarson**
## Change description

Previously, it was possible for two exports to be created for the same Record to a single Destination, resulting in multiple entries in the Destination. 

This PR adds a Redis lock during Export on a Record/Destination pairing.  When a batch is being sent, we now have the following logic:
1. For every export given: 
    i. Ensure it is the newest export for a record/destination pair.  If not, cancel it and look at the next given export.
    ii. Try to get a lock for the record/destination pair for the given export.  If no lock could be obtained, there is already an export in progress for that pair.  Do nothing, it will get tried again later.  If a lock was obtained, push it to an array of all current locks, push the export to an array of exports to be sent.
2. No change: If there are exports to be sent still, gather the needed methods and options and send them.
3. Finally, unlock all locks.

Also introduces a new type of lock `getLock`, which either returns an unlocking function if a lock was created or `null` if a lock was already in place for the given key.

### Development

- [x] Application changes have been tested appropriately

### Impact

- [x] Code follows company security practices and guidelines
- [x] Security impact of change has been considered
- [x] Performance impact of change has been considered
- [x] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [x] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

### CODE REVIEW

**Comment by teallarson** [comment]
@evantahler @pedroslopez this is still a WIP (needs testing)... before I dive into that, would like any feedback on the implementation now that I've got it working and existing tests passing!

File: plugins/@grouparoo/spec-helper/src/lib/specHelper.ts [hunk 1]
```diff
@@ -124,7 +124,7 @@ export namespace helper {
   export const longTime = 60 * 1000; // 60 seconds
   export const mediumTime = 20 * 1000; // 20 seconds
   export const shortTime = 10 * 1000; // 10 seconds
-  export const defaultTime = process.env.CI ? mediumTime : shortTime;
+  export const defaultTime = process.env.CI ? mediumTime : longTime;
```
Comment (evantahler) [inline plugins/@grouparoo/spec-helper/src/lib/specHelper.ts]: Why the change?  You can tell a specific test of describe to have a longer timeout if needed

File: core/src/modules/locks.ts [hunk 2]
```diff
@@ -41,3 +41,35 @@ export async function waitForLock(
 
   return { releaseLock, attempts };
 }
+
+export type getLockResponse = {
+  releaseLock: Function;
+  isLocked: boolean;
+  lockedBy: string;
+};
```
Comment (evantahler) [inline core/src/modules/locks.ts]: Since you aren't re-using this type anywhere, it's usually better to just have the getLock method determine it's own type from the implementation.   This way, there's nothing to change downsteam

File: core/src/modules/locks.ts [hunk 3]
```diff
@@ -41,3 +41,35 @@ export async function waitForLock(
...
+
+export async function getLock(
+  key: string,
+  requestId: string = uuid.v4(),
+  ttl = LOCK_DURATION_MS,
+  lockType?: string
```
Comment (evantahler) [inline core/src/modules/locks.ts]: for pattern matching, it might be best to give this a default, like `lockType = 'lock'`, this way all the lock keys will have the same number of "parts", eg: `grouparoo:lock:{type}:key` if we need to find or clear them in the future

File: core/src/modules/locks.ts [hunk 4]
```diff
@@ -41,3 +41,35 @@ export async function waitForLock(
...
+): Promise<getLockResponse> {
+  const client = api.redis.clients.client;
+  const lockKey = `grouparoo:lock:${
+    lockType.length > 1 && `${lockType}:`
+  }${key}`;
+
+  const set = await client.setnx(lockKey, requestId);
+
+  await client.expire(lockKey, Math.ceil(ttl / 1000));
```
Comment (evantahler) [inline core/src/modules/locks.ts]: This changes the expire TTL on the key regardless of this caller created it.  I would guess you don't want to mess with someone else's lock unless set===true

File: core/src/modules/locks.ts [hunk 5]
```diff
@@ -41,3 +41,35 @@ export async function waitForLock(
...
+
+  const isLocked = !set;
+
+  async function releaseLock() {
+    await client.del(lockKey);
+  }
```
Comment (evantahler) [inline core/src/modules/locks.ts]: Similar to the above, perhaps if you didn't create the lock, you shouldn't have access to the releaseLock method

File: core/src/modules/locks.ts [hunk 6]
```diff
@@ -41,3 +41,35 @@ export async function waitForLock(
...
+
+  const lockValue = await client.get(lockKey);
```
Comment (evantahler) [inline core/src/modules/locks.ts]: Is the lock value meaningful for the caller?  Perhaps I only care if I was able to get the lock or not

File: core/src/modules/ops/destination.ts [hunk 7]
```diff
@@ -912,13 +913,41 @@ export namespace DestinationOps {
...
   }> {
-    const _exports: Export[] = []; // only ones we are sending
+    const _exports: Export[] = [];
+
     for (const _export of givenExports) {
-      // TODO: maybe should recalcuate hasChanges and from current mostRecent
-      if (!_export.hasChanges) {
-        await _export.complete(); // do not include exports with hasChanges=false
+      const { isLocked, lockedBy } = await getLock(
+        `${_export.recordId}:${_export.destinationId}`,
```
Comment (evantahler) [inline core/src/modules/ops/destination.ts]: 👍 the lock key is a tuple comprised of the record.id and destination.id

File: core/src/modules/ops/destination.ts [hunk 8]
```diff
@@ -912,13 +913,41 @@ export namespace DestinationOps {
...
+            state: "canceled",
+            sendAt: null,
+            errorMessage: `Outdated duplicate of export ${lockingExport.id}`,
+            errorLevel: null,
+            completedAt: new Date(),
+          });
+          return;
+        } else if (lockingExport.id !== _export.id) {
+          //if the export we're looking at didn't create this lock, keep it pending for now
+          return;
```
Comment (evantahler) [inline core/src/modules/ops/destination.ts]: This return would exit the whole function. perhaps `continue` instead to go to the next _export in the loop

File: core/src/modules/ops/destination.ts [hunk 9]
```diff
@@ -981,6 +1010,18 @@ export namespace DestinationOps {
...
     }
 
+    for (const _export of _exports) {
+      // release locks for all of the exports that were processed regardless of outcome - re-enqueues are handled elsewhere
+      const { releaseLock } = await getLock(
+        `${_export.recordId}:${_export.destinationId}`,
+        _export.id,
+        null,
+        "export"
+      );
```
Comment (evantahler) [inline core/src/modules/ops/destination.ts]: Ah, I see while you are already returning `releaseLock` now.  Perhaps above in the first instance of getLock, you could store all the `releaseLock`s you get in an array and `Promise.all(locks.map(releaseLock => releaseLock())`.  This way, you don't need to talk to redis again

Comment (evantahler) [reply core/src/modules/ops/destination.ts]: It's also probably a good idea to wrap most of this method in a try/finally block so you can still release the locks if something goes wrong

File: core/src/modules/ops/destination.ts (hunk 7 above)
Comment (evantahler) [reply core/src/modules/ops/destination.ts]: I think `lockedBy` is a random UUID, so won't the `await Export.findById(lockedBy);` below always fail to find an Export?  I'll add an idea below to not need store any information in the lock

File: core/src/modules/ops/destination.ts [hunk 10]
```diff
@@ -912,13 +913,41 @@ export namespace DestinationOps {
...
-        await _export.complete(); // do not include exports with hasChanges=false
+      const { isLocked, lockedBy } = await getLock(
+        `${_export.recordId}:${_export.destinationId}`,
+        _export.id,
+        null,
+        "export"
+      );
+
+      if (isLocked) {
+        const lockingExport = await Export.findById(lockedBy);
```
Comment (evantahler) [inline core/src/modules/ops/destination.ts]: Perhaps this should be 

```ts
const mostRecentExport = await Export.findOne({where: {recordId: export.recordId, destinationId: export.destinationId}, order: ['createdAt', 'ASC']})
```

Then, you can be sure you have the most recent export and you can compare if `mostRecentExport.id === _export.id`.  Either "I" am the most recent export or not

File: core/src/modules/ops/destination.ts [hunk 11]
```diff
@@ -912,13 +913,41 @@ export namespace DestinationOps {
...
+        //if the export we're looking at is older than the one that locked this pair, cancel it
+        if (lockingExport.createdAt > _export.createdAt) {
+          await _export.update({
+            state: "canceled",
+            sendAt: null,
+            errorMessage: `Outdated duplicate of export ${lockingExport.id}`,
+            errorLevel: null,
+            completedAt: new Date(),
+          });
+          return;
```
Comment (evantahler) [inline core/src/modules/ops/destination.ts]: Same note about swapping `return` for `continue`

**Review by evantahler — COMMENTED** [review]
Comments commented!

File: core/src/modules/locks.ts (hunk 5 above)
Comment (bleonard) [reply core/src/modules/locks.ts]: Yeah, maybe we call this `getLockIfAvailable` or something. It would return `null` if it's already locked and the `releaseLock` if we were able to get it.

File: core/src/modules/ops/destination.ts [hunk 12]
```diff
@@ -912,13 +913,45 @@ export namespace DestinationOps {
...
+        const mostRecentExport = await Export.findOne({
+          where: {
+            recordId: givenExport.recordId,
+            destinationId: givenExport.destinationId,
+            state: "pending",
+          },
+          order: ["createdAt", "ASC"],
+        });
+        //if the export we're looking at is older than the one that locked this pair, cancel it
+        if (mostRecentExport.createdAt > givenExport.createdAt) {
```
Comment (bleonard) [inline core/src/modules/ops/destination.ts]: So if I read this right, we couldn't get the lock (because an export is being processsed in another thread or earlier in this batch) _and_ now we check against other pending exports. That right?

If the `mostRecentExport` is newer, I understand cancelling this one and `continue`.
But if `givenExport` is newer, we will still want to try it later. I think it should still `continue` though because letting it fall through like it is now means we still end up with them in parallel. right?

File: core/src/modules/ops/destination.ts [hunk 13]
```diff
@@ -912,13 +913,45 @@ export namespace DestinationOps {
...
+
+    for (const givenExport of givenExports) {
+      const lock = await getOrSetLock({
+        key: `${givenExport.recordId}:${givenExport.destinationId}`,
+      });
+      locks.push(lock);
+
+      if (lock.isLocked) {
+        // is _export the newest export for this pair?  if not, cancel it.
+        const mostRecentExport = await Export.findOne({
```
Comment (bleonard) [inline core/src/modules/ops/destination.ts:935]: Note this could, and mostly commonly would, return `givenExport`.
Maybe it should say `not` that one and then check for null.

File: core/src/modules/ops/destination.ts [hunk 14]
```diff
@@ -981,6 +1014,10 @@ export namespace DestinationOps {
       await app.checkAndUpdateParallelism("decr");
     }
 
+    for (const lock of locks) {
```
Comment (bleonard) [inline core/src/modules/ops/destination.ts]: put this in a `finally` so that if it crashes, we will still release them.

File: core/src/modules/ops/destination.ts (hunk 12 above)
Comment (teallarson) [reply core/src/modules/ops/destination.ts]: I did what you suggested below, checking for a `moreRecentExport` (the same query as before, but excluding the current export by id) and if one exists, cancelling the one we're looking at.  I think that should address the case you're talking about here, yes?

File: core/src/modules/ops/destination.ts (hunk 9 above)
Comment (tealjulia) [reply core/src/modules/ops/destination.ts]: I've done that with locks... implementing the try/finally block sounds right but the scoping of it is escaping me.  There are a handful of loops happening here, any of which could fail.

File: core/src/modules/ops/destination.ts (hunk 12 above)
Comment (bleonard) [reply core/src/modules/ops/destination.ts]: But if `givenExport` is the most recent and `lock.isLocked`, doesn't that mean that we are currently processing an older export that got the lock? In that case, we should wait to process `givenExport` until later. As it is now, it still gets processed in parallel with the one that got the lock.

It's all very tricky, so let me know if I'm wrong.

File: core/src/modules/ops/destination.ts [hunk 15]
```diff
@@ -912,13 +913,45 @@ export namespace DestinationOps {
...
+            errorMessage: `Replaced by more recent export ${mostRecentExport.id}`,
+            errorLevel: null,
+            completedAt: new Date(),
+          });
+          continue;
+        }
+      }
+
+      //things are locked, make sure we're using the absolute most recent data.
+      const updatedExport = await Export.findById(givenExport.id);
```
Comment (bleonard) [inline core/src/modules/ops/destination.ts]: how can this be updated? What changes do you expect?

File: core/src/modules/locks.ts (hunk 5 above)
Comment (tealjulia) [reply core/src/modules/locks.ts:70]: I think that makes sense.  During the unlocking in this particular instance, it doesn't much matter who created the lock.  The locks are just stored in an array and processed one-by-one after export.  But in terms of overall design, I think it makes sense to only give the `releaseLock()` out once.

File: core/src/modules/ops/destination.ts (hunk 15 above)
Comment (teallarson) [reply core/src/modules/ops/destination.ts]: This makes sure we're using the most recent properties.  `oldProperties` and `newProperties` in the export are populated with getters/setters on column read.

File: core/src/modules/ops/destination.ts (hunk 12 above)
Comment (teallarson) [reply core/src/modules/ops/destination.ts]: Nope that sounds right.  I had a `continue` in there earlier for that case...  It should have stayed.

File: core/src/modules/ops/destination.ts (hunk 10 above)
Comment (tealjulia) [reply core/src/modules/ops/destination.ts]: I'm wondering if we want `startedAt` or `createdAt` here... `startedAt` has to do with the batch it's in I believe.  In testing, I'm seeing two exports may have the same `startedAt` while they have differing `createdAt`s.

File: core/src/modules/ops/destination.ts [hunk 16]
```diff
@@ -912,82 +912,124 @@ export namespace DestinationOps {
...
+            moreRecentExport.id !== givenExport.id
+          ) {
+            await givenExport.update({
+              state: "canceled",
+              sendAt: null,
+              errorMessage: `Replaced by more recent export ${moreRecentExport.id}`,
+              errorLevel: null,
+              completedAt: new Date(),
+            });
+            continue;
```
Comment (bleonard) [inline core/src/modules/ops/destination.ts]: this one is now extra

Comment (teallarson) [reply core/src/modules/ops/destination.ts]: oh yep!

File: core/src/modules/ops/destination.ts (hunk 15 above)
Comment (bleonard) [reply core/src/modules/ops/destination.ts]: really? that's fascinating. You're saying by reading `newProperties`, it actually does a `SELECT FROM recordProperties` and not just reading it from what is set in the column? -- that seems surprising to me

I see this:
```
  get newRecordProperties(): ExportRecordProperties {
    return ExportOps.deserializeExportRecordProperties(
      //@ts-ignore
      this.getDataValue("newRecordProperties")
    );
  }
```
But that doesn't refresh the Record/Properties as far as I can tell.

Comment (teallarson) [reply core/src/modules/ops/destination.ts]: You're right.  It calls that function on each read, but that function isn't where they're actually gathered from.  On it.

Comment (bleonard) [reply core/src/modules/ops/destination.ts]: I'm thinking for now we don't actually have to update them. The "newest" checks you do above means it should work out.

Comment (bleonard) [reply core/src/modules/ops/destination.ts]: So I think you can just use `givenExport` below here.

Comment (bleonard) [reply core/src/modules/ops/destination.ts]: Well.... it's tricky. we are only doing that when you _don't_ get the lock now.
I'd say pair with @evantahler today and see where you end up.

Comment (teallarson) [reply core/src/modules/ops/destination.ts]: I think we ended up in a good direction, though I'm still seeing some test failures remotely, so it may need adjusting.

File: core/src/modules/ops/destination.ts [hunk 17]
```diff
@@ -912,82 +923,109 @@ export namespace DestinationOps {
...
+          where: {
+            recordId: givenExport.recordId,
+            destinationId: destination.id,
+          },
+          order: [["createdAt", "DESC"]],
+        });
+
+        const isNewest = mostRecentExport.id === givenExport.id;
+        if (!isNewest) console.info(mostRecentExport);
+        const gotLock = typeof lock === "function";
```
Comment (bleonard) [inline core/src/modules/ops/destination.ts]: what else can it be?

File: core/src/modules/ops/destination.ts [hunk 18]
```diff
@@ -912,82 +923,109 @@ export namespace DestinationOps {
...
+
+        if (gotLock) locks.push(lock);
+        if (!isNewest) {
+          await cancelOldExport(givenExport, mostRecentExport);
```
Comment (bleonard) [inline core/src/modules/ops/destination.ts:945]: I like canceling either way. Maybe do this before getting the lock if we are going to do it either way and bail. One less thing to unlock.

File: core/src/modules/ops/destination.ts [hunk 19]
```diff
@@ -912,82 +923,109 @@ export namespace DestinationOps {
...
+          continue;
+        }
+        if (gotLock && isNewest && givenExport.hasChanges)
+          _exports.push(givenExport);
+      }
```
Comment (bleonard) [inline core/src/modules/ops/destination.ts:960]: maybe return early here if `_exports.length === 0` - it's now possible.

File: core/src/modules/ops/destination.ts [hunk 20]
```diff
@@ -912,82 +923,109 @@ export namespace DestinationOps {
...
-    }
+    const _exports: Export[] = [];
+    const locks: Awaited<ReturnType<typeof getLock>>[] = [];
 
-    const exportRecords: ExportRecordsPluginMethod = await getBatchFunction(
-      destination
-    );
+    try {
+      for (const givenExport of givenExports) {
+        if (!givenExport.hasChanges) await givenExport.complete();
```
Comment (bleonard) [inline core/src/modules/ops/destination.ts]: probably just `continue` here. Weird to also cancel it lower or something.
Then can remove the `hasChanges` check below as well.

File: core/src/modules/ops/destination.ts (hunk 17 above)
Comment (teallarson) [reply core/src/modules/ops/destination.ts]: `null`

Comment (bleonard) [reply core/src/modules/ops/destination.ts]: I meant `typeof` - I was just thinking we would use `!lock`

**Review by bleonard — APPROVED** [review]

File: core/src/modules/ops/destination.ts (hunk 17 above)
Comment (teallarson) [reply core/src/modules/ops/destination.ts]: For sure.  @evantahler had recommended this, I believe as a slightly tighter guarantee.

Comment (bleonard) [reply core/src/modules/ops/destination.ts]: As you like. I think it's even weirder if it somehow returns something else and we don't unlock it. I'd rather see the crash in `Promise.all` below if it's not a function somehow. But not strong feelings here.

File: core/src/modules/ops/destination.ts (hunk 18 above)
Comment (evantahler) [reply core/src/modules/ops/destination.ts:945]: 👍 agree - might as well always cancel old exports.

File: core/src/modules/ops/destination.ts [hunk 21]
```diff
@@ -912,82 +922,115 @@ export namespace DestinationOps {
...
+        await app.checkAndUpdateParallelism("decr");
+      }
+      return updateExports(
+        destination,
+        _exports,
+        exportResult,
+        combinedError,
+        synchronous
+      );
     } finally {
```
Comment (evantahler) [inline core/src/modules/ops/destination.ts:1031]: Theses nested try/catch/finally areas are a little worrysome to me.  Is there a away that we can only have one try/catch here?  The finally's both are doing cleanup work so it's possible to both release all the locks and fixup the parallelism in one go.  Another variable will be needed to keep track of if we should `app.checkAndUpdateParallelism("decr");`, but I think that's worth if to avoid the nesting

File: core/__tests__/models/destination/plugins/exportRecord.ts [hunk 22]
```diff
@@ -996,13 +996,13 @@ describe("models/destination - with custom exportRecord plugin", () => {
...
       expect(_export.retryCount).toBe(0);
 
-      // when the response is back to success
       exportProfileResponse = {
         success: true,
         error: undefined,
         retryDelay: undefined,
       };
 
+      foundSendTasks = await specHelper.findEnqueuedTasks("export:send");
```
Comment (evantahler) [inline core/__tests__/models/destination/plugins/exportRecord.ts]: Did you intend to assert something about foundSendTasks?

File: core/src/modules/ops/destination.ts [hunk 23]
```diff
@@ -912,82 +922,115 @@ export namespace DestinationOps {
...
+        if (!givenExport.hasChanges) {
+          await givenExport.complete();
+          continue;
+        }
 
-    const syncMode = await destination.getSyncMode();
-    const syncOperations: DestinationSyncOperations = syncMode
-      ? DestinationSyncModeData[syncMode].operations
-      : DestinationSyncModeData.sync.operations; // if destination does not support sync modes, allow all
+        const mostRecentExport = await Export.findOne({
```
Comment (evantahler) [inline core/src/modules/ops/destination.ts:935]: Calling out that this will, in the short term, slow down this method because we need to SELECT from exports 100 times now (once for each export in the batch).  Once we do https://www.pivotaltracker.com/story/show/181115564, we can skip this lookup

**Review by evantahler — APPROVED** [review]
Well done! There was some tricky logic here.

I think there's one more option to clarify what's going on with try/catches, but not required.

File: core/src/modules/ops/destination.ts (hunk 21 above)
Comment (teallarson) [reply core/src/modules/ops/destination.ts:1031]: The `finally`s can be cleaned up fairly simply.  It seems like we're depending on things throwing in that "inside" try/catch though ([example](https://github.com/grouparoo/grouparoo/blob/04570d22a1772065d3ca17f288cc8b1904bef7b1/core/__tests__/models/destination/plugins/exportRecords.ts#L866))  I'm sure there's a way to get them unnested.  I have not found it yet though.
